### PR TITLE
refactor: WIP: Use an explicit type to define `Gas`

### LINKED ISSUE #8221: Consider using an explicit type for `Gas`
https://github.com/near/nearcore/blob/01c6f6f8ed6f4bbd15ece896d728145f6f993be5/core/primitives-core/src/types.rs#L26

We currently define `Gas` as an alias to `u64`.  This means that throughout the code, we can use the two types interchangeably and also perform arbitrary arithmetic on `Gas` values.  This issue is to discuss if it will make more sense to introduce a new type for `Gas` with some associated method: e.g. 

```Rust
struct Gas(u64);
impl Gas {
  pub fn checked_add(...) -> Result<...> {...}
  pub fn saturating_add(...) {...}
  ...
}
```

The above definitions can make the code more verbose but potentially more readable and reduce the chances of accidentally ignoring arithmetic overflows, etc. (e.g.

### PR DESCRIPTION
**Pull request by akhi3030**
Closes #8221 
Still WIP

### CODE REVIEW

File: core/primitives-core/src/types.rs [hunk 1]
```diff
@@ -23,7 +29,77 @@ pub type ShardId = u64;
...
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct Gas(u64);
+
+impl Gas {
+    pub fn new(val: u64) -> Self {
+        Self(val)
+    }
```
Comment (nagisa) [inline core/primitives-core/src/types.rs:52]: Consider `impl From<u64> for Gas`

File: core/primitives-core/src/types.rs [hunk 2]
```diff
@@ -23,7 +29,77 @@ pub type ShardId = u64;
...
+        self.0.checked_sub(rhs.0).map(Self::new)
+    }
+
+    pub fn wrapping_sub(self, rhs: Self) -> Self {
+        Self(self.0.wrapping_sub(rhs.0))
+    }
+
+    pub fn get(&self) -> u64 {
+        self.0
+    }
```
Comment (nagisa) [inline core/primitives-core/src/types.rs:80]: This should probably not exist. Looking at the uses, there are couple common situations:

* Divide gas into portions (`impl ops::Div<u64> for Gas { type Result = Gas; }` and same for `Rem`);
  * or add an inherent method...
* Divide gas by gas (`impl ops::Div<Gas> for Gas { type Result = u64; }`);
  * or add an inherent method, note though that the result is unitless.
* Subtraction/addition/etc – use one of the methods defined above. Same for the multiplication and such;
* (Parts of) complex computations like in `compute_new_gas_price` seems like could be inherent methods on the type.

File: core/primitives-core/src/types.rs [hunk 3]
```diff
@@ -23,7 +29,77 @@ pub type ShardId = u64;
...
+    Deserialize,
+    Clone,
+    Copy,
+    Hash,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct Gas(u64);
```
Comment (nagisa) [inline core/primitives-core/src/types.rs:47]:
```suggestion
#[repr(transparent)]
pub struct Gas(u64);
```

File: core/primitives-core/src/types.rs [hunk 4]
```diff
@@ -23,7 +29,77 @@ pub type ShardId = u64;
...
 pub type Balance = u128;
 /// Gas is a type for storing amount of gas.
-pub type Gas = u64;
+#[derive(
+    Debug,
+    Default,
+    BorshSerialize,
+    BorshDeserialize,
+    Serialize,
+    Deserialize,
```
Comment (nagisa) [inline core/primitives-core/src/types.rs:38]: These may need additional attributes to make sure that the serialized structure shape doesn't change.

Also 
```suggestion
    BorshSerialize,
    BorshDeserialize,
    serde::Serialize,
    serde::Deserialize,
```

as per the [style guide](https://github.com/near/nearcore/blob/master/docs/practices/style.md#derives).

File: core/primitives-core/src/runtime/fees.rs [hunk 5]
```diff
@@ -29,19 +29,23 @@ impl Fee {
...
+            self.send_sir.clone()
         } else {
-            self.send_not_sir
+            self.send_not_sir.clone()
         }
     }
 
     pub fn exec_fee(&self) -> Gas {
-        self.execution
+        self.execution.clone()
```
Comment (nagisa) [inline core/primitives-core/src/runtime/fees.rs:39]: `Gas` is `Copy` already, this is not necessary.

File: core/primitives-core/src/types.rs (hunk 1 above)
Comment (akhi3030) [reply core/primitives-core/src/types.rs:52]: Good idea!  I suppose that will also allow me to get rid of `get()`?  Also do you have any rule of thumbs or links that can help one decide when to use `new` vs `impl From`

Comment (akhi3030) [reply core/primitives-core/src/types.rs:52]: I found https://www.reddit.com/r/rust/comments/k7z66j/should_i_define_new_or_implement_from_for/.

Comment (nagisa) [reply core/primitives-core/src/types.rs:52]: >  Also do you have any rule of thumbs or links that can help one decide when to use new vs impl From

For newtypes that represent dimensional data (i.e. units) from/into my intuition is that this is more similar to conversion from `u32` to `u64`, than anything else which have plenty of rules of thumbs and conventions built around `From`.

OTOH, there are plenty of cases where `From`/`Into` is just impossible to use sensibly (e.g. the type needs two values to construct it…) too, so an inherent constructor method is the way to go there.

Usually I just go with the gut though.

File: core/primitives-core/src/types.rs (hunk 2 above)
Comment (akhi3030) [reply core/primitives-core/src/types.rs:76]: Good point.  Agreed.  My current plan is to get the code base compiling using this method for now and then see what the common patterns are that can be factored out.

**Comment by akhi3030** [comment]
I believe this PR is no longer actively being worked on.

**Pull request closed without merging**